### PR TITLE
Add debug checking of the loop table

### PR DESCRIPTION
**Pull request by SingleAccretion**
Partially addresses #8944.

I decided to extract this into its own function as only a handful of phases maintain the invariants it checks right now.

These are some issues/`TODO`s that I've encountered/noticed while working on this:

There are at least three places where where the `fgDebugCheck*` methods are called right now:
  - `phase.cpp`, where for some select phases most of the checks are called.
  - `Compiler::compCompile`.
  - Various phase methods themselves.

Comments in `phase.cpp` indicate that it's "the place" for placing these debug checks going forward, but I found it difficult to fit the new method there, as it introduces a check that only a few phases guarantee and those phases are not in the allow-list. I would love to clean this all up and replace with a system where phases declare their pre-conditions and post-conditions, all consolidated in `phase.cpp`. I've left this for a future PR.

Removing loops from the table is not done consistently: `optUnrollLoops` nulls-out some fields "for good measure", `optUpdateLoopsBeforeRemoveBlock` doesn't, neither does `fgFoldConditional`. At the same time `optUpdateLoopsBeforeRemoveBlock` does not update the "needs alignment" flag for the first block, while the two other methods do. It would be nice to consolidate this code into one function. It is not trivial though, because of the following issue.

`bbNatLoopNum` is not maintained well for blocks. For example, `optUnrollLoops` leaves behind some blocks with it set after it unrolls. This is closely related to the above problem and complicated by the fact that if we remove the inner loop, some blocks could presumably become part of the outer loop, and I suspect that there are a lot of edge cases to the logic that will have to be written to account for that, so I excluded `optUnrollLoops` from the checks in this PR.

Logging is not performed consistently between `#ifdef DEBUG` + `printf`s and `JITDUMP`s. I've cleaned this up a little for `fgdiagnostic.cpp`, but decided to leave a much larger commit for `optimizer.cpp` out. Also, functions headers in `fgdiagnostic.cpp` do not conform to the standard pattern.

This change survives the diff of the framework with crossgen (no results as expected), but I expect there will be CI failures to resolve. I may be a little slow to respond in the next couple of days.

Edit: given the CI is green'ish, marking this ready for review.

cc @BruceForstall

### CODE REVIEW

File: src/coreclr/jit/fgdiagnostic.cpp [hunk 1]
```diff
@@ -1945,12 +1935,7 @@ static volatile int bbTraverseLabel = 1;
 
 void Compiler::fgDebugCheckBBlist(bool checkBBNum /* = false */, bool checkBBRefs /* = true  */)
 {
-#ifdef DEBUG
-    if (verbose)
-    {
-        printf("*************** In fgDebugCheckBBlist\n");
-    }
-#endif // DEBUG
+    JITDUMP("*************** In fgDebugCheckBBlist\n");
```
Comment (BruceForstall) [inline src/coreclr/jit/fgdiagnostic.cpp:1938]: This is generally a fine conversion, to reduce code size and visible ifdefs. However, there is a trade-off: having the "printf" line allows for an easy place to put a breakpoint that only triggers on functions where you've set `COMPlus_JitDump` (so `verbose` is true), so I prefer to leave the original form here for "starting a 'phase'" kind of output like this.

**Review by BruceForstall — COMMENTED** [review]
I'm going to trigger pri-1 tests

File: src/coreclr/jit/fgdiagnostic.cpp (hunk 1 above)
Comment (SingleAccretion) [reply src/coreclr/jit/fgdiagnostic.cpp]: That is something I did not consider. Reverted.

**Comment by SingleAccretion** [comment]
@BruceForstall is anything more required from my side on this one?

**Review by BruceForstall — APPROVED** [review]
LGTM. Thanks for the contribution.

**Comment by BruceForstall** [comment]
> @BruceForstall is anything more required from my side on this one?

Nope... I was out for a few days...

**Pull request merged**